### PR TITLE
planner/cascades: introduce preprocessing phase in cascades planner

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

This PR introduces `preprocessing` phase in cascades planner.

After investigating the implementation of Apache Calcite, some optimization rules like `Column Pruning` might not be efficient to implement in cascades planner. Besides, rules like `Build KeyInfo` are also better to be implemented in the preprocessing phase. So this PR introduce a preprocessing phase in the cascades planner. 

### What is changed and how it works?

Cascades planner contained two phases (`exploration, implementation`) before, but after this PR, it will be composed of three phases: `preprocessing` -> `exploration` -> `implementation`.

Currently, we only have `ColumnPruning` and `BuildKeyInfo` in the preprocessing phase.

### Check List 

Side effects
 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

**Comment by winoros** [comment]
`KeyInfo` should finally be a property maintained everytime when a plan operator is generated. Not a rule.

**Comment by francis0407** [comment]
Yeah, we need to build the key info during the exploration for the new generated exprs. But I think, anyway, we have to build the key info for the first initialized groups before the exploration. If we don't build it here, how about build it in the function `convert2Group`? @winoros

File: planner/cascades/optimize.go [hunk 1]
```diff
@@ -67,10 +67,21 @@ func (opt *Optimizer) GetImplementationRules(node plannercore.LogicalPlan) []Imp
...
-// optimization is composed of 2 phases: exploration and implementation.
+// optimization is composed of 3 phases: preprocessing, exploration and implementation.
 //
 //------------------------------------------------------------------------------
-// Phase 1: Exploration
+// Phase 1: Preprocessing
+//------------------------------------------------------------------------------
+//
+// The target of this phase is to preprocess the logical plan tree by some heuristic
+// rules which may traverse the whole plan tree.
```
Comment (zz-jason) [inline planner/cascades/optimize.go]: // The target of this phase is to preprocess the plan tree by some heuristic
// rules which should always be beneficial, for example `Column Pruning`.

Comment (zz-jason) [reply planner/cascades/optimize.go]: With this comment, we can remove the next comment line.

Comment (francis0407) [reply planner/cascades/optimize.go]: addressed

**Comment by winoros** [comment]
@francis0407 
I prefer move it to `convert2GroupExpr` and attach it to LogicalProperty. But yes it's something we can do it later.
Add a TODO there to indicate that this method is not complete would lgtm.

**Pull request merged**